kernel/src/actions/mod.rs

```diff
@@ -138,7 +143,10 @@ impl Protocol {
     pub fn try_new_from_data(data: &dyn EngineData) -> DeltaResult<Option<Protocol>> {
         let mut visitor = ProtocolVisitor::default();
         data.extract(get_log_schema().project(&[PROTOCOL_NAME])?, &mut visitor)?;
-        Ok(visitor.protocol)
+        visitor
+            .protocol
+            .map(UnvalidatedProtocol::validate)
+            .transpose()
     }
 
```

Comment:
snapshot/transaction objects: check read/write features.

Reply:
flag: things that mutate state on filesystem without actual commit (log cleanup/VACUUM/etc.)

Reply:
`protocol.is_read_supported()` and write